src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -217,29 +219,33 @@ public void ExtractToFile(string destinationFileName, bool overwrite)
-        // /// <para>-or-</para>
-        // /// <para>An I/O problem occurred.</para></exception>
-        // /// <exception cref="InvalidOperationException">Attempted to extract an unsupported entry type.</exception>
-        // /// <exception cref="UnauthorizedAccessException">Operation not permitted due to insufficient permissions.</exception>
-        // public Task ExtractToFileAsync(string destinationFileName, bool overwrite, CancellationToken cancellationToken = default)
-        // {
-        //     throw new NotImplementedException();
-        // }
+        /// <summary>
+        /// Asynchronously extracts the current entry to the filesystem.
+        /// </summary>
+        /// <param name="destinationFileName">The path to the destination file.</param>
+        /// <param name="overwrite"><see langword="true"/> if this method should overwrite any existing filesystem object located in the <paramref name="destinationFileName"/> path; <see langword="false"/> to prevent overwriting.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None" />.</param>
+        /// <returns>A task that represents the asynchronous extraction operation.</returns>
+        /// <remarks><para>Files of type <see cref="TarEntryType.BlockDevice"/>, <see cref="TarEntryType.CharacterDevice"/> or <see cref="TarEntryType.Fifo"/> can only be extracted in Unix platforms.</para>
+        /// <para>Elevation is required to extract a <see cref="TarEntryType.BlockDevice"/> or <see cref="TarEntryType.CharacterDevice"/> to disk.</para></remarks>
+        /// <exception cref="ArgumentException"><paramref name="destinationFileName"/> is <see langword="null"/> or empty.</exception>
+        /// <exception cref="IOException"><para>The parent directory of <paramref name="destinationFileName"/> does not exist.</para>
+        /// <para>-or-</para>
+        /// <para><paramref name="overwrite"/> is <see langword="false"/> and a file already exists in <paramref name="destinationFileName"/>.</para>
+        /// <para>-or-</para>
+        /// <para>A directory exists with the same name as <paramref name="destinationFileName"/>.</para>
+        /// <para>-or-</para>
+        /// <para>An I/O problem occurred.</para></exception>
+        /// <exception cref="InvalidOperationException">Attempted to extract an unsupported entry type.</exception>
+        /// <exception cref="UnauthorizedAccessException">Operation not permitted due to insufficient permissions.</exception>
+        public Task ExtractToFileAsync(string destinationFileName, bool overwrite, CancellationToken cancellationToken = default)
+        {
+            if (EntryType is TarEntryType.SymbolicLink or TarEntryType.HardLink)
+            {
+                throw new InvalidOperationException(string.Format(SR.TarEntryTypeNotSupportedForExtracting, EntryType));
+            }
```

Comment:
Typically any exceptions other than argument exceptions are wrapped in the returned task, e.g.
```C#
return Task.FromException(new InvalidOperationException(string.Format(SR.TarEntryTypeNotSupportedForExtracting, EntryType));
```

Reply:
Will do.

Why do we not wrap argument exceptions?

Reply:
This feels like a good candidate for a Roslyn analyzer, if we don't have one yet.

Reply:
> Why do we not wrap argument exceptions?

Because in typical usage they represent a 100% deterministic developer error in how the API is being used.  We don't want such errors being eaten by the returned task, especially if doing so could lead to a deadlock, e.g. if this task won't end up performing some work that would ultimately unblock something else.

> And what about ArgumentNullException?

That is an argument exception.  So same deal.

Reply:
Thanks for the explanation.

I opened this API suggestion in case it makes sense to add such analyzer: https://github.com/dotnet/runtime/issues/70905

Reply:
@stephentoub does it make sense to add async methods all the way down to Interop.IOErrors.cs? I noticed we throw exceptions there, but on async code they would not be thrown properly as described above.

I'm thinking of adding methods like this one:

```cs
    internal static Task<int> CheckIoAsync(int result, string? path = null, bool isDirectory = false, Func<ErrorInfo, ErrorInfo>? errorRewriter = null) =>
        result < 0 ?
            Task.FromException<int>(GetExceptionForIoErrno(Sys.GetLastErrorInfo(), path, isDirectory, errorRewriter)) :
            Task.FromResult(result);
```

Reply:
Can you give me an example of what wouldn't be handled properly?

Reply:
I have this method in TarEntry.Unix.cs:
```cs
        // Unix specific implementation of the method that extracts the current entry as a hard link.
        private void ExtractAsHardLink(string targetFilePath, string hardLinkFilePath)
        {
            Debug.Assert(EntryType is TarEntryType.HardLink);
            Debug.Assert(!string.IsNullOrEmpty(targetFilePath));
            Debug.Assert(!string.IsNullOrEmpty(hardLinkFilePath));
            Interop.CheckIo(Interop.Sys.Link(targetFilePath, hardLinkFilePath), hardLinkFilePath);
        }
```

And I want to write an appropriate async version of this method. The `Interop.CheckIo` method is supposed to throw if the p/invoke fails. So I was thinking of adding a method like the one I shared in my previous comment, and then call it like this in the new async method:

```cs
        // Unix specific implementation of the method that asynchronously extracts the current entry as a hard link.
        private Task ExtractAsHardLinkAsync(string targetFilePath, string hardLinkFilePath)
        {
            Debug.Assert(EntryType is TarEntryType.HardLink);
            Debug.Assert(!string.IsNullOrEmpty(targetFilePath));
            Debug.Assert(!string.IsNullOrEmpty(hardLinkFilePath));
            return Interop.CheckIoAsync(Interop.Sys.Link(targetFilePath, hardLinkFilePath), hardLinkFilePath);
        }
```

Are there scenarios where I don't need to wrap the exception with `[Value]Task.FromException`?

Reply:
I know there isn't one, but my question was around handling the exception.

Reply:
There's no reason to create asynchronous version of ExtractAsHardLinkAsync or CheckIo, so I'm not clear on your question. Can you clarify?

Reply:
So the async method that calls `ExtractAsHardLink` should wrap that in a try catch, and the catch should return `Task.FromException`?

```cs
internal Task MethodAsync(...)
{
try
{
    ExtractAsHardLink(...);
    return Task.CompletedTask;
}
catch (Exception e)
{
    return Task.FromException(e);
}
}
```

Reply:
In other words, I should only create a new async method if it will call another async method inside it.

But in the case of methods that can exceptions, the last async method in the callstack is the one which should handle the exception by catching it and return it inside a Task.

Is this correct?

Reply:
What really matters is the public function black-box experience.

In the call of

```C#
Task t = someObjectOrType.FooAsync(arg0, arg1, arg2);
t.Wait();
Console.WriteLine(t.Exception);
```

There are two possibilities for the exception:

1) the call to FooAsync was invalid, in which case the FooAsync method itself should throw, and we never make it to line 2.  (ArgumentException)
2) We created a Task. The task ran... until it failed.  The exception is now in the t.Exception property.

For the latter, this is accomplished in one of three ways.

I) After argument validation, FooAsync failed synchronously, and the implementation returns `Task.FromException(e)`.
II) You're building tasks by chaining them together with Task.ContinueWith(action), inside the action you throw an exception (as per normal).  The Task system records the exception and changes the task state to Faulted.  https://github.com/dotnet/runtime/blob/988f570dc4ddd79a166ca8cbc184b2938ed2fefc/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L2332-L2357
III) You're writing a method with `async`/`await`.  That's just a fancy way of doing II.

What that boils down to is:

* In the public method you might need to be aware of being "inside" or "outside" the task with an exception, here's where Task.FromException comes in.
* Once you enter an `async`/`await` method you can stop thinking, just write throw as normal.
* If you're making Tasks by hand, and you're inside a callback, throw as normal.
* If you're making Tasks by hand and you're not inside a callback, Task.FromException may again matter.
  * But why are you making Tasks by hand after the public entrypoint?